packages/utils/dev-dep-resolver/src/lib.rs

```diff
@@ -0,0 +1,593 @@
+use std::{
+  borrow::Cow,
+  path::{Component, Path, PathBuf},
+};
+
+use dashmap::{DashMap, DashSet};
+use es_module_lexer::{lex, ImportKind};
+use parcel_resolver::{
+  CacheCow, FileSystem, Invalidations, ModuleType, Resolution, ResolveOptions, Resolver,
+  ResolverError, Specifier, SpecifierError, SpecifierType,
+};
+// use rayon::prelude::{ParallelBridge, ParallelIterator};
+
+#[derive(Debug)]
+pub enum EsmGraphBuilderError {
+  IOError(std::io::Error),
+  ParseError,
+  ResolverError(ResolverError),
+  Dynamic,
+  PatternError(glob::PatternError),
+  GlobError(glob::GlobError),
+  SpecifierError(SpecifierError),
+}
+
+impl From<std::io::Error> for EsmGraphBuilderError {
+  fn from(e: std::io::Error) -> Self {
+    EsmGraphBuilderError::IOError(e)
+  }
+}
+
+impl From<usize> for EsmGraphBuilderError {
+  fn from(_: usize) -> Self {
+    EsmGraphBuilderError::ParseError
+  }
+}
+
+impl From<ResolverError> for EsmGraphBuilderError {
+  fn from(e: ResolverError) -> Self {
+    EsmGraphBuilderError::ResolverError(e)
+  }
+}
+
+impl From<glob::PatternError> for EsmGraphBuilderError {
+  fn from(value: glob::PatternError) -> Self {
+    EsmGraphBuilderError::PatternError(value)
+  }
+}
+
+impl From<glob::GlobError> for EsmGraphBuilderError {
+  fn from(value: glob::GlobError) -> Self {
+    EsmGraphBuilderError::GlobError(value)
+  }
+}
+
+impl From<SpecifierError> for EsmGraphBuilderError {
+  fn from(value: SpecifierError) -> Self {
+    EsmGraphBuilderError::SpecifierError(value)
+  }
+}
+
+#[derive(Default)]
+pub struct Cache {
+  entries: DashMap<PathBuf, Invalidations>,
+}
+
+struct EsmGraphBuilder<'a, Fs> {
+  visited: DashSet<PathBuf>,
+  visited_globs: DashSet<PathBuf>,
+  invalidations: Invalidations,
+  cjs_resolver: Resolver<'a, Fs>,
+  esm_resolver: Resolver<'a, Fs>,
+  cache: &'a Cache,
+}
+
+impl<'a, Fs: FileSystem> EsmGraphBuilder<'a, Fs> {
+  pub fn build(&self, file: &Path) -> Result<(), EsmGraphBuilderError> {
+    if self.visited.contains(file) {
+      return Ok(());
+    }
+
+    self.visited.insert(file.to_owned());
+
+    if let Some(ext) = file.extension() {
+      if ext != "js" && ext != "cjs" && ext != "mjs" {
+        // Ignore.
+        return Ok(());
+      }
+    }
+
+    if let Some(invalidations) = self.cache.entries.get(file) {
+      self.invalidations.extend(&invalidations);
+      for p in invalidations.invalidate_on_file_change.iter() {
+        self.build(&p)?;
+      }
+      return Ok(());
+    }
+
+    let invalidations = Invalidations::default();
+    let module_type = self
+      .esm_resolver
+      .resolve_module_type(file, &invalidations)?;
+    let resolver = match module_type {
+      ModuleType::CommonJs | ModuleType::Json => &self.cjs_resolver,
+      ModuleType::Module => &self.esm_resolver,
+    };
+    let contents = resolver.cache.fs.read_to_string(&file)?;
+    let module = lex(&contents)?;
+    module
+      .imports()
+      // .par_bridge()
```

Comment:
I want to make this run in parallel, but that only works if the file system is the native one and not provided by JS otherwise we'll deadlock. Could work if we make the `packageManager.getInvalidations` API async but maybe that's a breaking change.